### PR TITLE
Feat/coinjoin final

### PR DESCRIPTION
**Pull request by szymonlesisz**
The missing part of https://github.com/trezor/trezor-suite/pull/6485

### Real coinjoin round process implementation

Commits are ordered by phase:
1. Round selection
> Pick right set of utxos to coordinator status rounds using coinjoin middleware binary
2. Input registration and connection confirmation
> Together because they are related to each other, input once registered needs to be constantly confirmed (more in code comments)
> @trezor/coinjoin requests for `ownership` is emitted from here
3. Output registration
> Join inputs credentials, calculate output amounts, split joined credentials and register outputs. ufff :)
4. Transaction signing
> Collect all inputs/outputs from coordinator status, sort, decide which are internal, temporary request affiliate server 
> @trezor/coinjoin requests for `signature` is emitted from here
> suite provides signatures in next commit
5. resolve @trezor/coinjoin requests in suite

### CODE REVIEW

File: packages/coinjoin/src/client/CoinjoinPrison.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+// Errored or currently registered inputs and addresses are sent here
+// inspiration: WalletWasabi/WabiSabi/Backend/Banning/Prison.cs
+
+export class CoinjoinPrison {
+    lastChange: number;
+    inmates: PrisonInmate[] = [];
+
+    constructor() {
+        this.lastChange = Date.now();
```
Comment (matejkriz) [inline packages/coinjoin/src/client/CoinjoinPrison.ts]: not used

File: packages/coinjoin/src/client/CoinjoinPrison.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        return this.inmates.find(i => i.id === id);
+    }
+
+    // called on each status change before rounds are processed
+    release() {
+        const now = Date.now();
+        this.inmates = this.inmates.filter(inmate => inmate.sentenceEnd > now);
+    }
+
+    amnesty(id: string) {
```
Comment (matejkriz) [inline packages/coinjoin/src/client/CoinjoinPrison.ts]: nit (naming): I think amnesty is usually for everyone so maybe `releaseById`(with `releaseBySentenceEnd`) or `releaseOne`?

File: packages/coinjoin/src/client/CoinjoinRound.ts [hunk 3]
```diff
@@ -114,41 +120,18 @@ export class CoinjoinRound extends EventEmitter implements SerializedCoinjoinRou
...
-        timeoutFn();
+        return selectRound(
+            (...args: ConstructorParameters<typeof CoinjoinRound>) => new CoinjoinRound(...args),
+            (...args: ConstructorParameters<typeof Alice>) => new Alice(...args),
+            accounts,
+            statusRounds,
+            coinjoinRounds,
+            prison,
+            options,
+        );
```
Comment (matejkriz) [inline packages/coinjoin/src/client/CoinjoinRound.ts:134]: nit:
would be a bit more readable if selectRound accepts params as one object so the parameters are named:
```suggestion
        return selectRound({
            roundGenerator: (...args: ConstructorParameters<typeof CoinjoinRound>) =>
                new CoinjoinRound(...args),
            aliceGenerator: (...args: ConstructorParameters<typeof Alice>) => new Alice(...args),
            accounts,
            statusRounds,
            currentRounds: coinjoinRounds,
            prison,
            options,
        });
```

File: packages/coinjoin/src/client/Alice.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+export class Alice {
+    path: string; // utxo derivation path
+    outpoint: string;
+    amount: number;
+    inputSize: number;
+    outputSize: number;
+    accountKey: string; // Account.accountKey
+    scriptType: AllowedScriptTypes; // Account.accountKey
```
Comment (mroz22) [inline packages/coinjoin/src/client/Alice.ts]: nitpick. maybe wrong comment. same for both lines? `// Account.accountKey`

File: packages/coinjoin/src/client/CoinjoinPrison.ts (hunk 2 above)
Comment (szymonlesisz) [reply packages/coinjoin/src/client/CoinjoinPrison.ts]: [5239ec1](https://github.com/trezor/trezor-suite/pull/6712/commits/5239ec1cee6e303565e09581d67ece666cff2b0a)

File: packages/coinjoin/src/client/CoinjoinPrison.ts (hunk 1 above)
Comment (szymonlesisz) [reply packages/coinjoin/src/client/CoinjoinPrison.ts]: [5239ec1](https://github.com/trezor/trezor-suite/pull/6712/commits/5239ec1cee6e303565e09581d67ece666cff2b0a)

File: packages/coinjoin/src/client/CoinjoinRound.ts [hunk 5]
```diff
@@ -198,18 +186,78 @@ export class CoinjoinRound extends EventEmitter implements SerializedCoinjoinRou
...
+                log(`Resolving ${type} request for <${i.outpoint}> with error. ${i.error}`);
+                input?.setError(new Error(i.error));
+            } else if ('ownershipProof' in i) {
+                log(`Resolving ${type} request for <${i.outpoint}>`);
+                input?.setOwnershipProof(i.ownershipProof);
+            } else if ('signature' in i) {
+                log(`Resolving ${type} request for <${i.outpoint}>`);
+                input?.setWitness(i.signature, i.index);
+            }
+        });
```
Comment (mroz22) [inline packages/coinjoin/src/client/CoinjoinRound.ts]: nitpick: would add something like `  if (!input) return;` and get rid of optional chaining operators

```suggestion
     inputs.forEach(i => {
            const input = this.inputs.find(a => a.outpoint === i.outpoint);

            if (!input) return;

            // reset request in input
            input.setRequest();
            if ('error' in i) {
                log(`Resolving ${type} request for <${i.outpoint}> with error. ${i.error}`);
                input.setError(new Error(i.error));
            } else if ('ownershipProof' in i) {
                log(`Resolving ${type} request for <${i.outpoint}>`);
                input.setOwnershipProof(i.ownershipProof);
            } else if ('signature' in i) {
                log(`Resolving ${type} request for <${i.outpoint}>`);
                input.setWitness(i.signature, i.index);
            }
        });
```

File: packages/coinjoin/src/client/CoinjoinClient.ts [hunk 6]
```diff
@@ -19,6 +20,7 @@ interface Events {
     round: CoinjoinRoundEvent;
     request: CoinjoinRequestEvent[];
     exception: string;
+    log: string;
 }
 
 export declare interface CoinjoinClient {
```
Comment (mroz22) [inline packages/coinjoin/src/client/CoinjoinClient.ts:26]: unreleated to this PR
I think we should create something like reuseable "TypedEventEmitter". I am using something similar in another PR - repetition. Followup issue? 

edit: also CoinjoinRound class implements something similar

File: packages/coinjoin/src/client/CoinjoinRound.ts [hunk 7]
```diff
@@ -198,18 +186,78 @@ export class CoinjoinRound extends EventEmitter implements SerializedCoinjoinRou
         return this;
     }
 
-    private processPhase(_accounts: Account[]) {
+    private processPhase(accounts: Account[], prison: CoinjoinPrison) {
         this.lock = createRoundLock(this.options.signal);
```
Comment (mroz22) [inline packages/coinjoin/src/client/CoinjoinRound.ts:188]: question and not part of this PR:
why `createRoundLock` method is not member of CoinjoinRound class? is there any specific reason?

File: packages/coinjoin/src/client/round/inputRegistration.ts [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+
+    // request ZeroCredentials from the middleware. use them in coordinator.inputRegistration
+    const zeroAmountCredentials = await middleware.getZeroCredentials(
+        round.amountCredentialIssuerParameters,
+        { signal, baseUrl: middlewareUrl },
+    );
+    const zeroVsizeCredentials = await middleware.getZeroCredentials(
+        round.vsizeCredentialIssuerParameters,
+        { signal, baseUrl: middlewareUrl },
+    );
```
Comment (mroz22) [inline packages/coinjoin/src/client/round/inputRegistration.ts:52]: nitpic, possible optimization. these two could be fired in parallel maybe?

File: packages/coinjoin/src/client/round/inputRegistration.ts [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+
+    // setup random delay for registration request. note that this may cause that the input will not be registered if phase change before expected deadline
```
Comment (mroz22) [inline packages/coinjoin/src/client/round/inputRegistration.ts]: nit: maybe add reason why. is it because of obfuscation of network communication?

File: packages/coinjoin/src/client/round/inputRegistration.ts [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+            zeroVsizeCredentials,
+            {
+                signal,
+                baseUrl: coordinatorUrl,
+                identity: input.outpoint,
+                delay,
+            },
+        )
+        .catch(error => {
+            options.log(
```
Comment (mroz22) [inline packages/coinjoin/src/client/round/inputRegistration.ts:78]: would it make sense to do some good old console.error here in order to see from sentry how often this happens?

File: packages/coinjoin/src/client/coordinator.ts [hunk 11]
```diff
@@ -136,19 +137,14 @@ export const transactionSignature = (
...
-        {
-            recipient_name: name,
-            outputs,
-        },
-        options,
-    );
+export const getAffiliateRequest = (tx: CoinjoinAffiliateTx, options: RequestOptions) =>
+    request<CoinjoinAffiliateRequest>('get_coinjoin_request_suite', tx, {
+        ...options,
+        baseUrl: 'https://coinjoin-affiliate.corp.sldev.cz/',
```
Comment (mroz22) [inline packages/coinjoin/src/client/coordinator.ts]: followup here?  this is likely not inteded for production?

File: packages/coinjoin/src/utils/coordinatorUtils.ts [hunk 12]
```diff
@@ -31,3 +42,119 @@ export const getScriptPubKeyFromAddress = (
 
     throw new Error('Unsupported scriptType');
 };
+
+// check WabiSabi.scriptPubKey format by OP and return AllowedScriptType
+export const getScriptTypeFromScriptPubKey = (scriptPubKey: string) => {
```
Comment (mroz22) [inline packages/coinjoin/src/utils/coordinatorUtils.ts]: nitpick: I'd verbosely annotate this function with AllowedScriptTypes. Could be more future proof 

```suggestion
export const getScriptTypeFromScriptPubKey = (scriptPubKey: string): AllowedScriptTypes => {
```

File: packages/coinjoin/src/client/CoinjoinPrison.ts [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+        this.inmates = this.inmates.filter(inmate => inmate.sentenceEnd > now);
+    }
+
+    amnestyById(id: string) {
+        this.inmates = this.inmates.filter(inmate => inmate.id !== id);
+    }
+
+    amnestyForAll() {
+        this.inmates = [];
+    }
```
Comment (mroz22) [inline packages/coinjoin/src/client/CoinjoinPrison.ts]: this is never used. I'd suggest replacing unused members with a comment, something like

```
// not implemented
// - amnestyForAll
```

of course it applies only if it is not going to be use in foreseeable future

**Review by mroz22 — COMMENTED** [review]
so far the first batch of nitpicks. will continue in the morning

File: packages/coinjoin/src/client/Alice.ts [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+        this.path = utxo.path;
+        this.outpoint = utxo.outpoint;
+        this.amount = utxo.amount;
+        this.inputSize = getInputSize(scriptType);
+        this.outputSize = getOutputSize(scriptType);
+    }
+
+    setError(error: Error) {
+        this.error = error;
+        return this;
```
Comment (matejkriz) [inline packages/coinjoin/src/client/Alice.ts]: why `return this;`?

File: packages/coinjoin/src/client/CoinjoinClient.ts [hunk 15]
```diff
@@ -140,10 +148,30 @@ export class CoinjoinClient extends EventEmitter {
...
+            if (match.length > 16) {
+                return `${match.substring(1, 9)}...${match.substring(
+                    match.length - 9,
+                    match.length - 1,
+                )}`;
+            }
+            return `[redacted]`;
+        });
+
+        this.emit('log', redacted);
```
Comment (matejkriz) [inline packages/coinjoin/src/client/CoinjoinClient.ts:165]: not important for this PR, but what about this?
```suggestion
        this.emit('log', isDevEnv ? message : redacted);
```
Or what is the point of putting those data in logs if they are always redacted?

File: packages/coinjoin/src/client/round/inputRegistration.ts [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+                ? Math.floor(roundParameters.coordinationFeeRate.rate * input.amount)
+                : 0;
+
+        const miningFee = Math.floor((input.inputSize * roundParameters.miningFeeRate) / 1000);
+        const amount = input.amount - coordinatorFee - miningFee;
+        const vsize = roundParameters.maxVsizeAllocationPerAlice - input.inputSize;
+
+        // use Credentials to get RealCredentials. use them in outputConfirmation
+        const realAmountCredentials = await middleware.getRealCredentials(
+            [amount, 0],
```
Comment (matejkriz) [inline packages/coinjoin/src/client/round/inputRegistration.ts:130]: what the `0` as second amount means here?

File: packages/coinjoin/src/client/round/outputDecomposition.ts [hunk 17]
```diff
@@ -0,0 +1,212 @@
...
+        };
+    }
+
+    for (let i = 1; i < inputs.length; i++) {
+        const current = inputs[i];
+        if (current.error) {
+            options.log(`Trying to join input with error. <${current.outpoint}> ${current.error}`);
+            throw current.error;
+        }
+        options.log(
```
Comment (matejkriz) [inline packages/coinjoin/src/client/round/outputDecomposition.ts:52]: there's no need to redact those values in log?

File: packages/coinjoin/src/client/round/outputDecomposition.ts [hunk 18]
```diff
@@ -0,0 +1,212 @@
...
+        );
+
+        // TODO: delay + identity (new identity for each loop)
+        const joinedIssuance = await coordinator.credentialIssuance(
+            round.id,
+            realAmountCredentials,
+            realVsizeCredentials,
+            zeroAmountCredentials,
+            zeroVsizeCredentials,
+            { signal, baseUrl: coordinatorUrl, identity: Date.now().toString(), delay: 0 },
```
Comment (matejkriz) [inline packages/coinjoin/src/client/round/outputDecomposition.ts]: why identity by time?

File: packages/coinjoin/src/client/round/outputDecomposition.ts [hunk 19]
```diff
@@ -0,0 +1,212 @@
...
+        outputSize,
+        availableVsize, // TODO: sum all available vsize and use chunks in output registration (increase nr of outputs)
+        internalAmounts,
+        externalAmounts,
+        { signal, baseUrl: middlewareUrl },
+    );
+    options.log(`Decompose amounts: ${outputAmounts.join('')}`);
+    return outputAmounts.map(amount => {
+        const miningFee = Math.floor((outputSize * roundParameters.miningFeeRate) / 1000);
+        const coordinatorFee = 0; // TODO: should be Math.floor(roundParameters.coordinationFeeRate.rate * amount);
```
Comment (matejkriz) [inline packages/coinjoin/src/client/round/outputDecomposition.ts]: why isn't this TODO resolved right away?

File: packages/coinjoin/src/client/round/outputDecomposition.ts [hunk 20]
```diff
@@ -0,0 +1,212 @@
...
+    const groupInputsByAccount = round.inputs.reduce((result, input) => {
+        if (!result[input.accountKey]) {
+            result[input.accountKey] = [];
+        }
+        if (!input.confirmedAmountCredentials || !input.confirmedVsizeCredentials) {
+            throw new Error(`Missing confirmed credentials for ${input.outpoint}`);
+        }
+        result[input.accountKey].push(input);
+        return result;
+    }, {} as Record<string, Alice[]>);
```
Comment (matejkriz) [inline packages/coinjoin/src/client/round/outputDecomposition.ts]: nit: I think this way of typing `reduce` is slightly better (you don't need the type assertion)
```suggestion
    const groupInputsByAccount = round.inputs.reduce<Record<string, Alice[]>>((result, input) => {
        if (!result[input.accountKey]) {
            result[input.accountKey] = [];
        }
        if (!input.confirmedAmountCredentials || !input.confirmedVsizeCredentials) {
            throw new Error(`Missing confirmed credentials for ${input.outpoint}`);
        }
        result[input.accountKey].push(input);
        return result;
    }, {});
```

File: packages/coinjoin/src/client/CoinjoinClient.ts (hunk 6 above)
Comment (szymonlesisz) [reply packages/coinjoin/src/client/CoinjoinClient.ts:26]: and CoinjoinBackend, and each Blockchain Link class :)
question is where this abstract should be placed? another package?

File: packages/coinjoin/src/client/CoinjoinRound.ts (hunk 7 above)
Comment (szymonlesisz) [reply packages/coinjoin/src/client/CoinjoinRound.ts:188]: i guess no reason, i had this witten as a regular js module before, without class.
class format comes later, function stayed :) 
it's basically another just variant of `@trezor/utils` deferred and i was probably thinking about some consolidation of those two

File: packages/coinjoin/src/client/round/inputRegistration.ts (hunk 8 above)
Comment (szymonlesisz) [reply packages/coinjoin/src/client/round/inputRegistration.ts:52]: doesnt really mater, middleware requests are fast (localhost)

File: packages/coinjoin/src/client/round/selectRound.ts [hunk 21]
```diff
@@ -0,0 +1,230 @@
...
+            }
+
+            log(`Found account candidate <${accountKey}> with ${utxos.length} inputs`);
+            return {
+                ...account,
+                utxos,
+            };
+        }
+
+        log(`Skipping candidate <${accountKey}>. Utxos ${utxos.length} of ${account.utxos.length}`);
```
Comment (matejkriz) [inline packages/coinjoin/src/client/round/selectRound.ts]: `utxos.length` will always be === 0 here, right? Is it intentional?

File: packages/coinjoin/src/client/round/inputRegistration.ts (hunk 9 above)
Comment (szymonlesisz) [reply packages/coinjoin/src/client/round/inputRegistration.ts]: you mean why random delay?

File: packages/coinjoin/src/client/coordinator.ts (hunk 11 above)
Comment (szymonlesisz) [reply packages/coinjoin/src/client/coordinator.ts]: yes, it needs to be changed before the merge. i dont have prod url yet

File: packages/coinjoin/src/client/CoinjoinPrison.ts (hunk 13 above)
Comment (szymonlesisz) [reply packages/coinjoin/src/client/CoinjoinPrison.ts]: i will remove them :) and put them back when i figure out why did i create them

File: packages/coinjoin/src/client/CoinjoinClient.ts (hunk 15 above)
Comment (szymonlesisz) [reply packages/coinjoin/src/client/CoinjoinClient.ts:165]: you can still see first/last chars of xpub/address from redacted message so its enough for debugging but not enough to consider it as sensitive data

File: packages/coinjoin/src/client/round/inputRegistration.ts (hunk 16 above)
Comment (szymonlesisz) [reply packages/coinjoin/src/client/round/inputRegistration.ts:130]: it's a nature of coordinator format.
RealCredentials[] are always "a pair" its useful when you want to join two Credentials (see outputDemposition) but in this case you want for fill (or in other words you have data) only for 1 Credential

File: packages/coinjoin/src/client/round/outputDecomposition.ts (hunk 18 above)
Comment (szymonlesisz) [reply packages/coinjoin/src/client/round/outputDecomposition.ts]: just random identity :) i will add comment

File: packages/coinjoin/src/client/round/outputDecomposition.ts (hunk 19 above)
Comment (szymonlesisz) [reply packages/coinjoin/src/client/round/outputDecomposition.ts]: because it's a coordinator/middleware bug we've found
it should be done like that but it isnt (will not work) i has come comment regarding that, maybe was lost during rebases.
i'll try to find it

File: packages/coinjoin/src/client/round/selectRound.ts (hunk 21 above)
Comment (szymonlesisz) [reply packages/coinjoin/src/client/round/selectRound.ts]: utxos.length yes but account.utxos.length doesnt have to

utxos are picked from account.utxos so this message will tell us how many was not picked (0 of 3)

File: packages/coinjoin/src/client/round/outputDecomposition.ts (hunk 17 above)
Comment (szymonlesisz) [reply packages/coinjoin/src/client/round/outputDecomposition.ts:52]: its only amounts, nothing sensitive

File: packages/suite/src/utils/wallet/coinjoinUtils.ts [hunk 22]
```diff
@@ -149,3 +152,76 @@ export const getEstimatedTimePerRound = (skipRounds: boolean) =>
     skipRounds
         ? ESTIMATED_HOURS_PER_ROUND_WITH_SKIPPING_ROUNDS
         : ESTIMATED_HOURS_PER_ROUND_WITHOUT_SKIPPING_ROUNDS;
+
+/**
+ * Transform @trezor/coinjoin CoinjoinRequestEvent.CoinjoinTransactionData to @trezor/connect signTransaction params
+ * Params are profiled byt account since multiple account can participate in one CoinjoinRound
```
Comment (matejkriz) [inline packages/suite/src/utils/wallet/coinjoinUtils.ts]: typo 
```suggestion
 * Params are profiled by account since multiple account can participate in one CoinjoinRound
```

File: packages/coinjoin/src/client/round/inputRegistration.ts (hunk 10 above)
Comment (szymonlesisz) [reply packages/coinjoin/src/client/round/inputRegistration.ts:78]: input registration errors are we can say allowed, nothing will happen if it fails here, input will try again in different round later

Comment (szymonlesisz) [reply packages/coinjoin/src/client/round/inputRegistration.ts:78]: maybe i will put console.error into critical phase instead

File: packages/coinjoin/src/client/round/inputRegistration.ts (hunk 16 above)
Comment (matejkriz) [reply packages/coinjoin/src/client/round/inputRegistration.ts:130]: thanks for explanation 👍

File: packages/coinjoin/src/client/round/outputDecomposition.ts (hunk 18 above)
Comment (matejkriz) [reply packages/coinjoin/src/client/round/outputDecomposition.ts]: is it random enough? what about some util getRandomIdentity (as a followup)

File: packages/coinjoin/src/client/CoinjoinRound.ts [hunk 23]
```diff
@@ -173,15 +156,20 @@ export class CoinjoinRound extends EventEmitter implements SerializedCoinjoinRou
...
-        await this.processPhase(accounts);
+        await this.processPhase(accounts, prison);
 
-        const [inputs, failed] = arrayPartition(this.inputs, input => !input.error);
+        const [inputs, failed] = arrayPartition(this.inputs, input => {
+            if (input.error) {
+                prison.detain(input.outpoint, { roundId: this.id, reason: input.error.message });
+            }
+            return !input.error;
+        });
```
Comment (marekrjpolak) [inline packages/coinjoin/src/client/CoinjoinRound.ts]: I'd prefer not to have side effects inside selector function, and it'll be shorter as well.
```suggestion
        const [inputs, failed] = arrayPartition(this.inputs, input => !input.error);
        failed.forEach(input =>
            prison.detain(input.outpoint, { roundId: this.id, reason: input.error?.message }),
        );
```

File: packages/coinjoin/src/client/CoinjoinRound.ts [hunk 24]
```diff
@@ -198,18 +186,78 @@ export class CoinjoinRound extends EventEmitter implements SerializedCoinjoinRou
...
+        if (this.phase === RoundPhase.OutputRegistration) {
+            return outputRegistration(this, accounts, prison, processOptions);
+        }
+        if (this.phase === RoundPhase.TransactionSigning) {
+            return transactionSigning(this, processOptions);
+        }
+        if (this.phase === RoundPhase.Ended) {
+            return Promise.resolve(this);
+        }
+        return Promise.resolve(this);
```
Comment (marekrjpolak) [inline packages/coinjoin/src/client/CoinjoinRound.ts]: This would look better as a `switch`. Also, should the `default` case do the same thing as `RoundPhase.Ended`? Shouldn't we somehow log that there's some unexpected round phase?
```suggestion
        switch (this.phase) {
            case RoundPhase.InputRegistration:
                return inputRegistration(this, prison, processOptions);
            case RoundPhase.ConnectionConfirmation:
                return connectionConfirmation(this, processOptions);
            case RoundPhase.OutputRegistration:
                return outputRegistration(this, accounts, prison, processOptions);
            case RoundPhase.TransactionSigning:
                return transactionSigning(this, processOptions);
            case RoundPhase.Ended:
            default:
                return Promise.resolve(this);
        }
```

File: packages/coinjoin/src/client/round/selectRound.ts [hunk 25]
```diff
@@ -0,0 +1,230 @@
...
+
+export const selectRound = async (
+    roundGenerator: CoinjoinRoundGenerator,
+    aliceGenerator: AliceGenerator,
+    accounts: Account[],
+    statusRounds: Round[],
+    currentRounds: CoinjoinRound[],
+    prison: CoinjoinPrison,
+    options: CoinjoinRoundOptions,
+): Promise<CoinjoinRound | void> => {
```
Comment (marekrjpolak) [inline packages/coinjoin/src/client/round/selectRound.ts]: It doesn't feel right to return "something or void", but I have no strong argument against it.

File: packages/coinjoin/src/client/Alice.ts (hunk 14 above)
Comment (szymonlesisz) [reply packages/coinjoin/src/client/Alice.ts]: some leftover, removed

File: packages/coinjoin/src/client/round/outputDecomposition.ts (hunk 18 above)
Comment (szymonlesisz) [reply packages/coinjoin/src/client/round/outputDecomposition.ts]: and what it will return, we just need some random string here, i can use some trezor/utils here

File: packages/coinjoin/src/client/round/inputRegistration.ts (hunk 9 above)
Comment (mroz22) [reply packages/coinjoin/src/client/round/inputRegistration.ts]: yes

File: packages/coinjoin/src/client/CoinjoinRound.ts (hunk 3 above)
Comment (matejkriz) [reply packages/coinjoin/src/client/CoinjoinRound.ts:134]: followup

**Comment by szymonlesisz** [comment]
forcepushed first batch of fixes

File: packages/coinjoin/src/client/round/selectRound.ts (hunk 25 above)
Comment (szymonlesisz) [reply packages/coinjoin/src/client/round/selectRound.ts]: tbh return type doesnt have to be here at all, ts will guess it: Promise<CoinjoinRound | undefined>

Comment (szymonlesisz) [reply packages/coinjoin/src/client/round/selectRound.ts]: [732481b](https://github.com/trezor/trezor-suite/pull/6712/commits/732481b70daf5567bd5b23c089c3b2a292284be5)

File: packages/coinjoin/src/client/CoinjoinRound.ts (hunk 24 above)
Comment (szymonlesisz) [reply packages/coinjoin/src/client/CoinjoinRound.ts]: [732481b](https://github.com/trezor/trezor-suite/pull/6712/commits/732481b70daf5567bd5b23c089c3b2a292284be5)

File: packages/coinjoin/src/client/CoinjoinRound.ts (hunk 23 above)
Comment (szymonlesisz) [reply packages/coinjoin/src/client/CoinjoinRound.ts]: [732481b](https://github.com/trezor/trezor-suite/pull/6712/commits/732481b70daf5567bd5b23c089c3b2a292284be5)

File: packages/suite/src/actions/wallet/coinjoinClientActions.ts [hunk 26]
```diff
@@ -67,14 +74,41 @@ const clientSessionCompleted = (accountKey: string) =>
...
+    } as const);
+
+const clientLog = (payload: string) =>
+    ({
+        type: COINJOIN.CLIENT_LOG,
+        payload,
+    } as const);
+
 export type CoinjoinClientAction =
     | ReturnType<typeof clientEnable>
```
Comment (komret) [inline packages/suite/src/actions/wallet/coinjoinClientActions.ts:105]: What should the `clientEnable` action do? There is no case for `CLIENT_ENABLE` in `coinjoinReducer`.

Comment (szymonlesisz) [reply packages/suite/src/actions/wallet/coinjoinClientActions.ts:105]: it supposed to set some loader in the ui butiw wasnt implemented at the end, it is there for 3 months already :)

**Review by mroz22 — APPROVED** [review]
looks ok, I have some degree of understanding whats going on here (could be better though), and it works on my machine. there are still bugs but they may be unrelated. so I am creating issues and we can solve them later. For example https://github.com/trezor/trezor-suite/issues/6739

image

File: packages/coinjoin/src/client/CoinjoinRound.ts (hunk 3 above)
Comment (szymonlesisz) [reply packages/coinjoin/src/client/CoinjoinRound.ts:134]: followup here https://github.com/trezor/trezor-suite/issues/6698

**Review by matejkriz — APPROVED** [review]
